### ISSUE TITLE
testStreamRead and testBufReader failures

### ISSUE THREAD

**Issue by auouymous** [minetest/minetest#3943]
Is this a real bug or just because the values were set to 53.53467f and tested as 53.534f?

```
Test assertion failed: readF1000(is) == 53.534f
    at test_serialization.cpp:305
[FAIL] testStreamRead - 0ms
Test assertion failed: buf.getF1000() == 53.534f
    at test_serialization.cpp:472
[FAIL] testBufReader - 0ms

```

Minetest 0.4.13-dev-f9a9038-dirty
gentoo linux
gcc (Gentoo 4.9.3 p1.3, pie-0.6.3) 4.9.3

**Comment by kwolekr** [comment]
It happens sometimes with different compilers.

**Comment by est31** [comment]
Its a bug in the compiler if the behaviour is defined and the compiler violates it. If the behaviour is undefined here then its a bug in minetest, either the serialisation layer or the testing code.

**Comment by est31** [comment]
And its far more likely that the bug is in minetest.

**Comment by auouymous** [comment]
The above failures happen with -O1, -O2 and -O3. They change to these failures with -O0:

```
Test assertion failed: readF1000(is) == 53.534f
    at test_serialization.cpp:305
[FAIL] testStreamRead - 0ms
Test assertion failed: !memcmp(&data[0], test_serialized_data, data.size())
    at test_serialization.cpp:355
[FAIL] testStreamWrite - 0ms
Test assertion failed: !memcmp(&buf[0], test_serialized_data, sizeof(test_serialized_data))
    at test_serialization.cpp:415
[FAIL] testVecPut - 0ms

```

**Comment by kwolekr** [comment]
This isn't so much a bug as it is a matter of expecting floating point math to be more accurate than it is.  Have you tried compiling with -fno-fast-math?

**Comment by PilzAdam** [comment]
Why do we even use `==` to compare floating point numbers in unit tests? Tests should always expect floating point numbers to be a bit off, so using something like `actual - expected < someSmallDelta` should be used instead.

**Comment by sofar** [comment]
should just make a nice macro like this:

#define FLOAT_ROUGHLY_EQUALS(f1, f2) ((f1) - (f2) <= FLOAT_MIN_VARIANCE)

**Comment by PilzAdam** [comment]
@sofar you would also need an `abs` around the subtraction.

**Comment by ShadowNinja** [comment]
IIRC @Zeno- added a function to make these checks, but I can't find it now.
`actual - expected < 0.001` is already used though, and that should fix this.

**Comment by Zeno-** [comment]
@ShadowNinja I did make a proper "is close to" function and it was a PR but it was never merged. The problem with .001 (or any other fixed value like that) is that as numbers get bigger then the more it becomes incorrect. So strictly speaking doing something like comparing to .0001 (or whatever) is not great but I think, from memory, that I closed the PR because the range/precision of FP numbers in minetest was not worth doing a correct "is close to" function.

**Comment by Zeno-** [comment]
Wait, we're compiling with fast-math enabled? That's kinda scary

**Comment by auouymous** [comment]
-fno-fast-math doesn't change the failures, probably because the code isn't compiled with -Ofast or -ffast-math.

**Comment by pgimeno** [comment]
I've investigated this a little. I've distilled this test case from the failed tests: https://paste.scratchbook.ch/view/69368d2b

According to the report of someone who ran into this issue, the floats being compared are (in hex):

    actual = 0x1.ac45a4p+5; expected = 0x1.ac45a2p+5

This person's machine was IEEE-754 compliant. Given that the exact result is actually below the expected value, and when rounded it gives the expected value, the actual result can't come from standard IEEE-754 division (which requires the result to be the exact value rounded), no matter the rounding mode.

With the test program linked above, I've been able to reproduce the failure in GCC 6.3.0 with `-funsafe-math-optimizations` and with `-O1` or above. When checking the compiled output, the compiler actually generated a multiplication by `0.001f` instead of a division by `1000.f`, which explains the issue.

What's puzzling is that @auouymous said that `-fno-fast-math` doesn't change the failures. I wonder if `src/util/serialize.cpp` was recompiled, since it's the one that contains the offending division.

**Comment by pgimeno** [comment]
I got the same failure in a release build, and I expect that will happen to everyone.

This patch fixes it for me:
```diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 8218d586c..3ff234621 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -722,7 +722,7 @@ else()
        if(CMAKE_SYSTEM_NAME MATCHES "(Darwin|FreeBSD)")
                set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Os")
        else()
-               set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -ffast-math -fomit-frame-pointer")
+               set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -ffast-math -fno-reciprocal-math -fomit-frame-pointer")
        endif(CMAKE_SYSTEM_NAME MATCHES "(Darwin|FreeBSD)")
        set(CMAKE_CXX_FLAGS_SEMIDEBUG "-g -O1 -Wall -Wabi ${WARNING_FLAGS} ${OTHER_FLAGS}")
        set(CMAKE_CXX_FLAGS_DEBUG "-g -O0 -Wall -Wabi ${WARNING_FLAGS} ${OTHER_FLAGS}")
```

**Comment by nerzhul** [comment]
this problem has been fixed in 5.0
